### ISSUE TITLE
BleachBit is not accessible on Windows

### ISSUE THREAD

**Issue by mhussaincov** [bleachbit/bleachbit#719]
hi there,
tryed the latest copy of bleachbit here on windows 10 with the screen reader nvda,
I launch the programme and there is no accessibility,
no feedback from my screen reader,
could this be fixed?
could you post the last functioning version of bleachbit that accessibility works? I will have to use that version untell a fix is released?
thanks for your help!
Majid

**Comment by Tobias-B-Besemer** [comment]
BB is based on the GTK+ framework.
BB uses GTK+ v3 since BB3.
Seems NVDA is not compatible with GTK+ at all!
http://nabble.nvda-project.org/GTK-with-NVDA-td39492.html
This postings are from 2015, but if no one have ported the ATK framework to Windows by now, this is still up-to-date!
And unfortunately this all is nothing that @az0 can change... :-/

Have you used BB in the past with this reader?
Because - as I wrote above - the GTK+ version have lately changed in BB...
So had you success with your reader with BB versions <3.0 ???

And one other question:
Can you try to borrow e.g. from a friend a screen reader that is not from NVDA and based on the ATK framework ???
Because - as I read it - GTK+ supports other frameworks for screen reader...

**Comment by Tobias-B-Besemer** [comment]
K, seems Andrew had this issue already one year before with BB2 and can't change anything on the situation right now.
https://www.bountysource.com/issues/62414343-bleachbit-accessibility-with-screen-reader-on-windows
But as far as I understand, screen reader support is much better in Linux and there it should work with the Linux port of BB, too!
But I guess I can't you suggest to switch to Linux... ;)

**Comment by mhussaincov** [comment]
bleachbit v2.x does not work also,
so there is no solution for this??
Majid

On 01/12/2019, Tobias B. Besemer <notifications@github.com> wrote:
> K, seems Andrew had this issue already one year before with BB2 and can't
> change anything on the situation right now.
> https://www.bountysource.com/issues/62414343-bleachbit-accessibility-with-screen-reader-on-windows
> But as far as I understand, screen reader support is much better in Linux
> and there it should work with the Linux port of BB, too!
> But I guess I can't you suggest to switch to Linux... ;)
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/bleachbit/bleachbit/issues/719#issuecomment-560108041


-- 
kind regards,
Majid Hussain

**Comment by Tobias-B-Besemer** [comment]
As I wrote above, you can only try a other screen reader or switch to Linux... :-/
Sorry !!!

**Comment by Tobias-B-Besemer** [comment]
OK, I took the time and did some research for it!

But lets link this issue first to the other issues with nearly the same prob...
https://github.com/bleachbit/bleachbit/issues/379
https://github.com/nvaccess/nvda/issues/3369

Here are some general info to GTK:
https://en.m.wikipedia.org/wiki/GTK
And there you can read:
> After GTK 2.24.10 and 3.6.4 Development of Windows with Installer was closed by Gnome. Installation of MSYS2 on Windows is a good way to use actual GTK.
> GTK 2.24.10 and 3.6.4 is available in Internet, but very buggy and limited against actual versions.
> A Version for Windows 64-bit is prepared by Tom Schoonjans with 2.24.32 (actual like Linux) and 3.22.30 (actual like Linux) available.
@az0, which exact version of GTK do you use ATM ???

I also found this site:
https://thenerdshow.com/GTK+.html
There you can read:
> Cross-compiling Atk
I guess it's the ATK we're talking about... ^^

Then I found this repo:
https://github.com/GNOME/atk

And some ATK docs here:
https://developer.gnome.org/accessibility-devel-guide/stable/dev-start-5.html.en
https://developer.gnome.org/accessibility-devel-guide/stable/gad-how-it-works.html.en
https://developer.gnome.org/accessibility-devel-guide/stable/gad-custom.html.en
Unfortunately there is nothing to read if this also fits for Windows...

...and then I found this:
https://developer.gnome.org/atk/2.28/

...but @mhussaincov, to be true: Even if this should work - and we can'z
t be sure off - I'm not dure if @az0 find in the next months time to work on... :-/

A other thing:
I read of the in Windows integrated screen reader "Narrator"... Have you tested it, already ??? Is this screen reader able to read GTK probs ???

**Comment by az0** [comment]
BleachBit 3.0 on Windows uses GTK 3.18.9, which was released March 2016. While this is a bit old, it's much newer than the GTK version that BleachBit version 2.2 had. On Linux, the GTK version varies.

I don't see a simple solution. I targeted this to version 4.2 to revisit after the dust settles from the big changes with BleachBit 4

**Comment by Tobias-B-Besemer** [comment]
Btw, @mhussaincov: I think a big help would be, if you can test BB3 with your screen reader on Linux (e.g. Ubuntu) and provid info, if the screen read is able to work with BB there.

**Comment by DataTriny** [comment]
Hello,
Today I just tested BleachBit because I had enough of another popular Windows cleaning software I won't name... As a visually impaired person, and an NVDA user, I can tell that the situation has not evolved and the whole GUI is not exposed to the screen reader. Apart from that, keep up the good work!

**Comment by tapper82** [comment]
This really needs sorting out. If a framework you are using is not accessible then you need to change how the application works. This is 2021 and to have a application that will not work with a screen reader is just not OK. If you are not shure how to fix it why don't you try and write a blog post asking for some help after all this application is open-source.

**Comment by manuelcortez** [comment]
Hi,
As a Linux and windows blind user I can identify the issue. The problem is just that GTK has never worked in Windows with the accessibility stack provided by microsoft (from where the screen reading software available for the OS reads information) because ATK is only functional in Linux systems. In short, there should be a kind of compatibility layer between ATK and UIA or whatever is the name of the current libraries offering a11y information in windows. If you somehow would compile aATK under windows, there's At-SPI (the lib that provides the info to screen readers) and the fact that there is no windows equivalent to At-SPI (In short, it provides info for Linux tools in the way Gnome has done it but this way is not the same than windows' way).
If someone would be willing to take the time to do something, this might be porting the GUI to some other toolkit which would be both cross platform and accessible, such as WX (which uses native controls provided by the OS, which are already accessible and bound to a11y toolkits), or QT 5 (which has a binding to every major platform with accessibility support). However this might not be feasible.
Another consideration might be to create a command line version of the application or expose a kind of API from where someone else might build a fork with accessibility included.

**Comment by abitrolly** [comment]
> Another consideration might be to create a command line version of the application or expose a kind of API from where someone else might build a fork with accessibility included.

`bleachbit.py` works without GUI if invoked from command line with arguments.

Yet another variant is to raise the issue through humanitarian channel to raise the problem with Microsoft adding accessibility layer to Gtk for their platform. They are already platinum members of Linux Foundation https://www.linuxfoundation.org/en/join/members/ and if they really want to change their reputation, this is something they should at least consider.

**Comment by az0** [comment]
My ideal solution is for someone, such as Microsoft, to add accessibility support to Gtk on Windows as @abitrolly suggested, but this is beyond the scope of this project. There is an article on gtk.org from October 2020 with a summary of [major accessibility changes in GTK 4](https://blog.gtk.org/2020/10/21/accessibility-in-gtk-4/), but it does not mention Windows.

As Anatoli also pointed out, BleachBit already has a command line version. 

I am open to exploring a new GUI for BleachBit in wxWidgets or QT, but such a change would not be a small or quick effort.

**Comment by tapper82** [comment]
Hi has any one found any more info out about this issue?

**Comment by KoroushGhazi** [comment]
To the Bleachbit developers:

Guys, I fully understand that changing the underpinnings of a program to make it more compatible with Windows in particular can be a pain. But now is the perfect time to pursue this.

I've been a computer enthusiast for over 30 years, I used to run TweakGuides.com. Even though I'd heard of Bleachbit, I didn't realise it was a great replacement for the CCleaner, I thought it was just for secure deletion. With Avast effectively killing off the aforementioned hugely popular utility, there's now massive potential demand for Bleachbit among Windows users. And with a rapidly ageing population around the world, and no sign of Linux being more than a niche OS, it would be a sound investment if you actively explored how to make it work with various Windows-based readers.

If necessary, crowdfunding can be used and I'm sure you'll find many people, including me, ready to promote it. Please consider!

Cheers,
Koroush Ghazi
@KoroushGhazi on Twitter

**Comment by az0** [comment]
I'd like to revisit this after releasing a fix for https://github.com/bleachbit/bleachbit/issues/848 . After that, a possible next steps are research and maybe a proof of concept with wxWidgets (wxPython). Based on some quick research, I noticed the documentation for[ accessibility in wxWidgets](https://www.wxwidgets.org/docs/tutorials/accessibility/) is dated 2003. If anyone is interested, another way to "pre-test" is to install other applications [already built on wxPython](https://wiki.wxpython.org/wxPythonPit%20Apps).

**Comment by audiogamer22** [comment]
I'm blind and after downloading and buying CCleaner, I thought I'd try BleachBit to see if it's any better, it's not. It's still not accessible and honestly even though CCleaner is paid for and can be a little clunky at times, it's what I'll be using for my future cleaning needs.

If something changes; I'll think about coming back to this program and checking it out but for now, CCleaner is the best option, in my oppinion, for blind windows users.

**Comment by az0** [comment]
Thank you for your patience. Here is a long update. In summary, I tested with modern GTK 3 on Windows 11 with Microsoft Narrator, but I am sad to write that is no improvement in accessibility. 

For people interested in this feature, I am interested to hear from you on a few points. 

First, which screen reader software do you use? I ask to guide future efforts in testing. 

Second, what is your opinion on the command line interface? The CLI is available now.

Third, what is your opinion on the idea of interacting with BleachBit through an AI chatbot? AI chatbots now have agentic tool calling abilities, and a bridge could be created between an AI and BleachBit. You could type commands and questions in natural language, and it could execute commands such as selecting options, setting preferences, running a preview, cleaning, and summarizing the results. You would need to set up either a local AI chatbot or a cloud-based chatbot. Local chatbots such as LMStudio and ollama are free of charge, but without a powerful GPU in your PC, there is a noticeable tradeoff in speed versus intelligence. Claude Desktop App is a cloud-based chatbots that is available for a $20/month subscription. For this to work, I would create an MCP server for BleachBit that replaces the GUI, and the chatbot connects to the BleachBit MCP server. Of course, the chatbot itself would need to be accessible. If the AI chatbot idea interests  you, please [create a new issue ticket](https://github.com/bleachbit/bleachbit/issues/new), so we can focus this issue ticket (issue 719) on traditional accessibility of the GUI.

Finally, here are some details about my recent testing. For years, BleachBit was stuck on an old GTK version 3.18.9, but for the upcoming BleachBit version 5.0 release, I updated it to GTK 3.24. This update was a major effort because GTK is difficult to build for Windows. First, I tested GTK 3.24.43 built by vcpkg and MSVC. I also tested GTK 3.24.49 built by MSYS2 and GCC. In both configurations, I confirmed that atk (the GNOME accessibility framework) was present. I tested the BleachBit GUI as is, and I also tested a stand-alone "hello world" GTK 3 application. After that, I added accessibility code to the stand-alone application. In all these tests, the accessibility was broken. In most cases, Windows Narrator recognized only the main title of the window. In a few cases, Narrator spoke the letters as I typed them.

On Linux, I briefly tested BleachBit with a GNOME tool called accerciser, and this confirmed that it exposes accessibility information.

Going forward, potential solutions for accessibility on Microsoft Windows are 

1. Using the existing CLI.
2. Investigating the accessibility in GTK 4 on Windows.
3. Creating a new BleachBit GUI using a completely other toolkit such as QT or wxWidgets.
4. Developing an integration with an AI chatbot.

**Comment by DataTriny** [comment]
Hi @az0,

GTK recently merged an [experimental AccessKit integration] that is enabled on Windows and macOS.

[AccessKit](https://github.com/AccessKit/accesskit) is a library that makes it easier for UI toolkits to provide cross-platform accessibility. It implements UI Automation on Windows, which means that any software integrating AccessKit will work with all screen readers and other assistive technologies. As long as you can grab a pre-built binary or compile GTK 4 yourself (you will need a Rust compiler to compile AccessKit), this will be transparent to you.

If you plan to migrate to GTK 4, you should get accessibility support basically for free once this is done.

**Comment by az0** [comment]
I am closing this issue to merge it with a prior issue ticket to consolidate the discussion. For further updates, please visit https://github.com/bleachbit/bleachbit/issues/631 and click on the button labeled subscribe. 

Let's continue the conversation there about GTK 4 AccessKit, the idea from @KoroushGhazi  crowdfunding, and other options. 

Thank you for your patience.
